File: src/go/ch01/1.4/l1/float_num.go
package main

import "math"

// AlmostEqual reports true if |x - y| <= max(rel*max(|x|,|y|), abs).
func AlmostEqual(x, y, rel, abs float64) bool {
	ax, ay := math.Abs(x), math.Abs(y)
	diff := math.Abs(x - y)
	tol := math.Max(rel*math.Max(ax, ay), abs)
	return diff <= tol
}

// KahanSum: compensated summation, great for small/medium slices.
func KahanSum(xs []float64) float64 {
	var s, c float64
	for _, x := range xs {
		y := x - c
		t := s + y
		c = (t - s) - y
		s = t
	}
	return s
}

// PairwiseSum: stable tree/pairwise summation (iterative).
func PairwiseSum(xs []float64) float64 {
	n := len(xs)
	if n == 0 {
		return 0
	}
	buf := make([]float64, n)
	copy(buf, xs)
	m := n
	for m > 1 {
		k := 0
		i := 0
		for i+1 < m {
			buf[k] = buf[i] + buf[i+1]
			k++
			i += 2
		}
		if i < m {
			buf[k] = buf[i]
			k++
		}
		m = k
	}
	return buf[0]
}

// MachineEpsilon returns the smallest eps with 1.0+eps != 1.0.
func MachineEpsilon() float64 {
	eps := 1.0
	for 1.0+eps/2.0 != 1.0 {
		eps /= 2.0
	}
	return eps
}

// ULPDiff returns the distance in representable steps between a and b.
func ULPDiff(a, b float64) uint64 {
	ua := math.Float64bits(a)
	ub := math.Float64bits(b)
	if int64(ua) < 0 {
		ua = 0x8000000000000000 - ua
	}
	if int64(ub) < 0 {
		ub = 0x8000000000000000 - ub
	}
	if ua > ub {
		return ua - ub
	}
	return ub - ua
}

File: src/erlang/ch01/1.4/l1/src/float_num.erl
%% Float utilities: almost_equal, Kahan & pairwise sum, machine epsilon, ULP distance, nextafter_up.
-module(float_num).
-export([almost_equal/3, kahan_sum/1, pairwise_sum/1, machine_epsilon/0, ulp_diff/2, nextafter_up/1]).

%% Robust float comparison: |X-Y| =< max(Rel*max(|X|,|Y|), Abs)
almost_equal(X, Y, Rel, Abs) ->
    Diff = abs(X - Y),
    Ax = abs(X), Ay = abs(Y),
    Tol = max(Rel * max(Ax, Ay), Abs),
    Diff =< Tol.

kahan_sum(List) ->
    kahan_sum_(List, 0.0, 0.0).
kahan_sum_([], S, _C) -> S;
kahan_sum_([X|Xs], S, C) ->
    Y = X - C,
    T = S + Y,
    C1 = (T - S) - Y,
    kahan_sum_(Xs, T, C1).

pairwise_sum([]) -> 0.0;
pairwise_sum([X]) -> X;
pairwise_sum(List) ->
    pairwise_sum(pair_once(List, [])).

pair_once([], Acc) -> lists:reverse(Acc);
pair_once([A], Acc) -> lists:reverse([A|Acc]);
pair_once([A,B|Rest], Acc) -> pair_once(Rest, [A+B|Acc]).

machine_epsilon() ->
    machine_epsilon_(1.0).
machine_epsilon_(Eps) ->
    case 1.0 + Eps / 2.0 of
        1.0 -> Eps;
        _   -> machine_epsilon_(Eps / 2.0)
    end.

%% ULP distance between two floats (64-bit IEEE754).
ulp_diff(A, B) ->
    UA = float_bits_u64(A),
    UB = float_bits_u64(B),
    MA = map_for_order(UA),
    MB = map_for_order(UB),
    abs(MA - MB).

%% Next representable float strictly greater than F.
nextafter_up(F) when is_float(F) ->
    Bits = float_bits_u64(F),
    SignBit = Bits band 16#8000000000000000,
    NextBits = case SignBit of
        0 -> Bits + 1;                      %% positive or +0
        _ -> Bits - 1                       %% negative or -0
    end,
    bits_to_float(NextBits).

%% --- bit helpers ---

float_bits_u64(F) ->
    %% big-endian canonical form
    <<Bits:64/unsigned-big-integer>> = <<F:64/float-big>>,
    Bits.

bits_to_float(Bits) ->
    <<F:64/float-big>> = <<Bits:64/unsigned-big-integer>>,
    F.

map_for_order(UnsignedBits) ->
    %% Convert to a lexicographically increasing integer space for ULP diff
    Sign = UnsignedBits band 16#8000000000000000,
    case Sign of
        0 -> UnsignedBits;
        _ -> 16#8000000000000000 - UnsignedBits
    end.

max(A,B) -> if A >= B -> A; true -> B end.
